Add internal configuration option to control before or after app delegation for libraries

tjwatson opened the pull request:
- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

tjwatson committed:
Add options for patch library and library delegation before app class path

Adds config attribute called libraryDelegation to the global
classloading config that has options beforeApp and afterApp where
afterApp is the default behavior today.  When using beforeApp
option the library delegation will happen before the application
class path and before parent delegation.  The library loader will
also be configured as parentLast delegation.

Adds a config attribute on classloader config to called
patchLibraryRef which is similar to privateLibraryRef.  Both
options take the library class path content and place it
on the local class path of the application.  A patchLibrary
will add the content to the beginning of the local class path.
A privateLibrary adds the content to the end of the local class path

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=2d33e26b-19f3-426d-a92f-0e9417d67abf) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=5fddde44-924a-480d-b6dd-40e3152c6678) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e1cd1002-fee9-4763-927a-ab9e564731a2) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4c464012-291e-40e2-aded-4f7057eb9131) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson committed:
Add tests for patch libraries

This tests that the order is correct for searching various patch libraries
and common libraries.  The test application contains 8 classes that
are also provided by the various libraries configured as patch or
common.  It confirms that the correct patch library can override the class
contained in the application and that no common library is able to
override the class contained in the application.

The assertion is done by checking the codesource location is using
the expected JAR or WAR file.

Additional assertions are done for resource lookups to validate
the correct resource is used for getResource and that the order
is correct when calling getReources.

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6f50676f-3782-4a74-b1c5-a570e6c5a310) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson committed:
Add tests for libraryDelegation configuration

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e35d4fc0-94e7-4876-987d-a9c2c0f5c8ba) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

rickettmwork commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=eee8cf12-df0c-4bad-a5bf-abcc617243a0) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=48dcedbf-786f-4065-9c7e-6c6fa3b83568) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=fdc14618-fa1f-4174-9ef7-03ac8e9a31d6) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ca88a347-da92-4d13-b4af-2888c28e038c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson committed:
Move SPI for patch libraries to be an internal

Avoiding new SPI for patch libraries until we finalize
the configuration attribute name and placement.

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=914bbd7c-cf65-4401-8165-ffdd3365976d) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/providers/Providers.java:
```diff
@@ -63,11 +80,11 @@ public static List<Library> getPrivateLibraries(ClassLoaderConfiguration config)
-        GetLibraries getLibraries = new GetLibraries(config.getId().getId());
-        return BlockingListMaker.defineList().waitFor(10, SECONDS).fetchElements(getLibraries).listenForElements(getLibraries).log(LOGGER).useKeys(privateLibraries).make();
+        GetLibraries getLibraries = new GetLibraries(ownerId);
+        return BlockingListMaker.defineList().waitFor(10, SECONDS).fetchElements(getLibraries).listenForElements(getLibraries).log(LOGGER).useKeys(libraryIds).make();
```
I can't tell whether this blocking list contains patch libraries.  I would expect it to.

tjwatson replied:
This is called from both `getPatchLibraries` and `getPrivateLibraries` above so the answer is that it depends on who calls it.  I do think this method `getLibraries` probably should be made private if nobody else is calling it.

dazavala replied:
Perfect.  I missed that `getLibraries()` was called from `getPatchLibraries()`.  +1 on changing access to private.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ContainerClassLoader.java:
```diff
@@ -1239,7 +1292,7 @@ private static class SmartClassPathImpl implements SmartClassPath {
         }
 
-        final List<UniversalContainer> classPath = new CopyOnWriteArrayList<UniversalContainer>();
+        final UniversalContainerList classPath = new UniversalContainerList(true);
```
Given the change to use `copyOnWrite` list for class path entries, I looked and don't see how an instance of this list could be exported and become stale. If you see that possibility, please modify the solution to prevent it.

tjwatson replied:
I kept it using copyOnWrite within `UniversalContainerList` because that is what was done before when handing this out.  There are complications because of how private and patch library references get added to the classpath lazily from com.ibm.ws.classloading.internal.AppClassLoader.lazyInit().  I didn't really want to do a deep dive to figure how to make this no longer use a `copyOnWrite` list.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java:
```diff
@@ -12,7 +12,9 @@
 package com.ibm.ws.classloading.internal;
 
-import static com.ibm.ws.classloading.internal.AppClassLoader.SearchLocation.DELEGATES;
+import static com.ibm.ws.classloading.configuration.GlobalClassloadingConfiguration.LibraryPrecedence.beforeApp;
```
Copyright ^^^.  I'm finding a few of these as I go.  Just look over your changes and update them.  No more comments.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/ClassLoaderConfigHelper.java:
```diff
@@ -16,6 +16,7 @@
 import static com.ibm.ws.classloading.ClassLoaderConfigHelper.Attribute.classProviderRef;
```
Copyright ^^^

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java:
```diff
@@ -59,23 +62,35 @@ public Class<?> loadClass(String name) throws ClassNotFoundException {
     public Enumeration<URL> getResources(String resName) throws IOException {
-        // search order: 1) my class path 2) parent loader
-        return super.findResources(resName).add(this.parent.getResources(resName));
+        // search order: 1)  my class path 2) parent loader
```
Comment is obsolete now.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java:
```diff
@@ -59,23 +62,35 @@ public Class<?> loadClass(String name) throws ClassNotFoundException {
-        return result == null ? this.parent.getResource(resName) : result;
+    public URL getResource(String name) {
+        URL result = findResourceCommonLibraryClassLoaders(name, beforeApp);
+        if (result == null) {
```
Appreciate the re-org of these methods. Will help debug and service.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ShadowClassLoader.java:
```diff
@@ -162,6 +165,26 @@ protected Class<?> loadClass(String className, boolean resolveClass, boolean onl
+        ClassNotFoundException lastException = null;
+        for (LibertyLoader delegate : delegates) {
+            try {
+                Class<?> result = delegate.loadClass(className, false, false, returnNull);
```
I never noticed this before.  ShadowClassLoader invokes delegate class loaders w/o `onlySearchSelf`.  Sounds like a bug.

tjwatson replied:
yeah, I noticed that also, but was not willing to address that here.  Mainly because I am not convinced we have great test coverage of the `ShadowClassLoader`.

dazavala replied:
JPA uses the `ShadowClassLoader`.  IIRC think the intent is to have a class loader define application classes besides the app class loader (shadowed loader).  And that implies the class path search behavior should be the same as AppClassLoader, because we expect the linkage of classes loaded by ShadowClassLoader to be resemble those loaded by shadowed AppClassLoader.  The change in delegation behavior was an oversight IMO.  Notice the comment mentions `CompoundClassLoader` -- application class loader type for another platform.  But CCL delegates using `onlySearchSelf`.  So, no definite conclusions here after looking further.

```
            if (service != null) {
                // Create a temporary classloader with the same class loading characteristics
                // as the input CompoundClassLoader for use by the persistence provider.
                ClassLoader loader = service.getShadowClassLoader(classLoader);
                ctx.ungetService(reference);
                return loader;
            }
```

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/providers/GetLibraryLoaders.java:
```diff
@@ -30,30 +31,33 @@
 
     @Override
-    public LibertyLoader fetch(String id) throws ElementNotReadyException, ElementNotValidException {
+    public Providers.LibraryInfo fetch(String id) throws ElementNotReadyException, ElementNotValidException {
```
Had you considered adding precedence support to LibertyLoader instead of Provider?

tjwatson replied:
This may be a misnamed class `LibraryInfo`.  Probably should be something like `LibertyLoaderPrecedence`.  I wanted separate object here that held the configuration of the loader delegation which was separate from the `LibertyLoader` instances.  While it may be possible to move that down to the `LibertyLoader` I had to complications with that approach:

1. I was not sure it would apply to all the loaders that extend `LibertyLoader`
2. These LibertyLoader instances are created independently of the the `classloading` configuration of `libraryPrecedence` configuration.

For now this is all internal, so we can iterate on improving it.  One complication is that we do NOT change the `libraryPrecedence` delegation for `classProvider` references.  If we did then we wouldn't need this `LibraryInfo` class at all because we could assume all are using the same `libraryPrecedence` as the globally configured setting.

dazavala replied:
Not that I care to do this.  One idea to be use `libraryPrecedence` as a global setting:  Have the app CL create  equivalence sets (ordered, of course) of delegates based on provider type rather than precedence.  The app CL would then use `libraryPrecedence` in load operations to predicate calls to new method `findClassDelegateLoaders(List<ACL> delegates, ...)`,  passing it the library or classProvider delegates as required. That's not as elegant as this solution, but it reduces or even obviates the need to propagate the `libraryPrecedence` setting for specific types of provider delegates.

dazavala commented on dev/com.ibm.ws.classloading_classpath_fat/publish/servers/patchLibWarTest/server.xml:
```diff
@@ -0,0 +1,33 @@
+
+    <include location="../fatTestPorts.xml"/>
+
+    <webApplication location="testPatchLib.war">
```
FYI: a patch library could be overridden/ineffective for an embedded web module in an enterprise app configured with `parentLast` class loading delegation.  We're not testing that scenario here, obviously, but I I thought you might like a reminder before finalizing the patch support design.

tjwatson replied:
I was ignoring any notion of making a library reference of any kind: `commonLibrary`, `privateLibrary`, `patchLibrary` to be specific to individual modules inside an EAR.  Are you suggesting we need to open that discussion up to allow library references to be associated with individual modules (e.g. WARs) inside an EAR?  If so, how would you suggest we make such a configuration reference?

dazavala replied:
That's the problem.  There is no class loading configuration specific to embedded modules.  The configuration would be a property on the class loader that correlates a patch library to a module name.

tjwatson replied:
`patchLibrary` is no different than `privateLibrary` in this regard.  Both get added to the classpath of the top-level application class loader, therefore a `patchLibrary` really only makes sense to "patch" classes that are on the class path of the top-level class loader.  That is the EAR's loader for an EAR deployment and the WAR's loader for a WAR deployment.  I could also mention springBoot, but that is similar to a WAR deployment with regards to the top-level application loader.

dazavala reviewed (changes requested):
I left a few in-line comments.  Not worried about tests for `parentLast` inconsistent class space for now.

Q: Is there a relationship between a patch library reference and common library precedence`?  

I would like to revisit the design for app patching.  If `beforeApp` is in effect, a patch library could be overridden by a common library.  That's unlikely as a patch is intended for the application class path, and common libraries must not have application dependencies.  But, a patch library could be overridden/ineffective for an embedded web module in an enterprise app configured with `parentLast` class loading delegation.  A patch library prefers default class loading configuration.

tjwatson commented:
> Q: Is there a relationship between a patch library reference and common library precedence`?
> 
> I would like to revisit the design for app patching. If `beforeApp` is in effect, a patch library could be overridden by a common library. That's unlikely as a patch is intended for the application class path, and common libraries must not have application dependencies. But, a patch library could be overridden/ineffective for an embedded web module in an enterprise app configured with `parentLast` class loading delegation. A patch library prefers default class loading configuration.

`patchLibrary` and and `libraryPrececence` are orthogonal.  They are related to each other the same way a `privateLibrary` reference is related to a `commonLibrary` reference.

The default `libraryPrecedence` for  `commonLibrary` delegation comes after the local classpath search so anything on the local classpath can "override" what is available from the `commonLibrary` reference.  This includes what is in that application as well as the `privateLibrary` references.

If the `libraryPrecedence` is set to `beforeApp` then `commonLibrary` delegation comes before the local classpath search so anything on the local classpath can be "overriden by what is available by the `commonLibrary` reference.  This includes `patchLibrary` references.  If a user needs to "patch" a `commonLibrary` then they would simply place the patch JAR as the first entry of the library configuration for the `commonLibrary`.  They would not try to patch a library class with a local classpath `privateLibrary` or `patchLibrary` reference.

tjwatson committed:
Rename LibraryInfo to LoaderInfo and update copyright dates

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=dd211943-290d-474d-8efd-36569492bb3f) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

dazavala reviewed (approved):
Code changes look good.  I'll check any unresolved conversations now.

dazavala commented:
>  If a user needs to "patch" a commonLibrary then they would simply place the patch JAR as the first entry of the library configuration for the commonLibrary. They would not try to patch a library class with a local classpath privateLibrary or patchLibrary reference.

This is where things get dodgy, because an application patch that has dependencies on application classes cannot deploy as a common library.  But otherwise, yea.

dazavala commented:
@tjwatson I've offered all follow-on thoughts to our conversations.  Good luck going forward.

Hit the conversations drop down button on the [Files Changed tab](https://github.com/OpenLiberty/open-liberty/pull/32184/files).
